File: tom_som_go_runtime/spec_meta_bridge.go
package somruntime

// spec_meta_bridge.go — bridge from the meta-JSON class graph (SpecModel) to
// the canonical metadata tree (SomMetaTree, DR1 §3.1), a faithful port of
// `tom_som_dart_runtime/lib/src/spec_meta_bridge.dart` (and the TypeScript
// `spec_meta_bridge.ts`).
//
// The generated facades (DR8) emit populated SomMetaTrees directly; every
// consumer that only has the exported spec-model meta-data (the conformance
// harness, tooling) builds its tree here. The expansion follows the same walk
// SpecReflection.Resolve performs — crucially, a node's SomMetaNode.SectionID
// is the **field-level** id exactly as exported (field.sectionId), so the
// tree's path grammar stays byte-compatible with the paths a SpecDocument is
// keyed by. (DR1 §2.2's "field id, else target-class id" resolution is applied
// at *export* time by the model exporter / DR8 generator, not re-derived
// here.)
//
// Recursive class references become terminal re-entry nodes
// (SomMetaNode.Recursive), mirroring how the generated chains terminate.

import "sort"

// slottedAnnotations holds annotation names that have a dedicated SomMetaNode
// slot; everything else is captured losslessly into SomMetaNode.Extra.
var slottedAnnotations = map[string]bool{
	"SectionId":            true,
	"SectionIdPattern":     true,
	"SerializationOrder":   true,
	"Min":                  true,
	"Unused":               true,
	"ContentType":          true,
	"ContentHelp":          true,
	"Comment":              true,
	"Form":                 true,
	"Document":             true,
	"MapsTo":               true,
	"DetailedIn":           true,
	"SecondLevelSectionId": true,
}

// BuildSomMetaTree builds the wired SomMetaTree for one document root of
// model.
//
// When rootType is "" the model's first root is used; otherwise the root is
// looked up via SpecModel.RootByType (returning its error for an unknown
// type). Children are ordered by @SerializationOrder (annotated members first,
// then declaration order), which is the sibling emission order of the
// hierarchical YAML format (DR1 §2.3).
func BuildSomMetaTree(model *SpecModel, rootType string) (*SomMetaTree, error) {
	var root *SpecRoot
	if rootType == "" {
		if len(model.Roots) > 0 {
			root = model.Roots[0]
		} else {
			r, err := model.RootByType("")
			if err != nil {
				return nil, err
			}
			root = r
		}
	} else {
		r, err := model.RootByType(rootType)
		if err != nil {
			return nil, err
		}
		root = r
	}
	cls := model.ClassNamed(root.Type)

	sectionID := root.SectionID
	docComment := root.Doc
	classDoc, mapsTo, detailedIn := "", "", ""
	classSectionID := ""
	var annotations []*SpecAnnotation
	var children []*SomMetaNode
	if cls != nil {
		if sectionID == "" {
			sectionID = cls.SectionID
		}
		if docComment == "" {
			docComment = cls.Doc
		}
		classDoc = cls.Doc
		classSectionID = cls.SectionID
		mapsTo = cls.MapsTo
		detailedIn = cls.DetailedIn
		annotations = cls.Annotations
		children = bridgeChildNodes(model, cls, map[string]bool{root.Type: true})
	}
	rootNode := &SomMetaNode{
		ClassName:       root.Type,
		SectionID:       sectionID,
		ClassSectionID:  classSectionID,
		Kind:            SomMetaKindSection,
		TypeName:        root.Type,
		DocComment:      docComment,
		ClassDocComment: classDoc,
		MapsTo:          mapsTo,
		DetailedIn:      detailedIn,
		Document: &SomDocMeta{
			Name:        root.Title,
			Description: root.Description,
			BasedOn:     bridgeBasedOn(cls),
		},
		SecondLevelIds: bridgeSecondLevelIds(annotations),
		Extra:          bridgeExtras(annotations),
		Children:       children,
	}
	return NewSomMetaTree(rootNode)
}

func bridgeBasedOn(cls *SpecClass) []string {
	if cls == nil {
		return []string{}
	}
	ann := cls.Annotation("Document")
	if ann == nil {
		return []string{}
	}
	raw, ok := ann.Argument("basedOn").([]interface{})
	if !ok {
		return []string{}
	}
	out := make([]string, 0, len(raw))
	for _, e := range raw {
		out = append(out, bridgeStr(e))
	}
	return out
}

// bridgeStr coerces an annotation argument value to its string form ("" for
// nil, matching the other ports' String(x || ”)).
func bridgeStr(v interface{}) string {
	switch t := v.(type) {
	case string:
		return t
	case int:
		return itoa(t)
	case float64:
		if t == float64(int(t)) {
			return itoa(int(t))
		}
		return ""
	case bool:
		if t {
			return "true"
		}
		return "false"
	default:
		return ""
	}
}

func bridgeChildNodes(model *SpecModel, cls *SpecClass, stack map[string]bool) []*SomMetaNode {
	const fallback = 1 << 30
	type indexed struct {
		index int
		field *SpecField
	}
	fields := make([]indexed, len(cls.Fields))
	for i, f := range cls.Fields {
		fields[i] = indexed{index: i, field: f}
	}
	sort.SliceStable(fields, func(a, b int) bool {
		oa, ob := fallback, fallback
		if fields[a].field.SerializationOrder != nil {
			oa = *fields[a].field.SerializationOrder
		}
		if fields[b].field.SerializationOrder != nil {
			ob = *fields[b].field.SerializationOrder
		}
		if oa != ob {
			return oa < ob
		}
		return fields[a].index < fields[b].index
	})
	out := make([]*SomMetaNode, 0, len(fields))
	for _, f := range fields {
		out = append(out, bridgeFieldNode(model, cls, f.field, stack))
	}
	return out
}

func bridgeFieldNode(
	model *SpecModel, owner *SpecClass, field *SpecField, stack map[string]bool,
) *SomMetaNode {
	kind := field.Kind // SpecFieldKind* values equal the SomMetaKind* values.
	var target *SpecClass
	recursive := false
	var children []*SomMetaNode
	var elementNode *SomMetaNode

	if field.Kind == SpecFieldKindComplex || field.Kind == SpecFieldKindSection {
		target = model.ClassNamed(field.Type)
		if target != nil {
			if stack[target.Name] {
				recursive = true
			} else {
				stack[target.Name] = true
				children = bridgeChildNodes(model, target, stack)
				delete(stack, target.Name)
			}
		}
	} else if field.Kind == SpecFieldKindList && field.ElementIsComplex {
		element := model.ClassNamed(field.ElementType)
		if element != nil {
			var elementChildren []*SomMetaNode
			elementRecursive := false
			if stack[element.Name] {
				elementRecursive = true
			} else {
				stack[element.Name] = true
				elementChildren = bridgeChildNodes(model, element, stack)
				delete(stack, element.Name)
			}
			elementNode = &SomMetaNode{
				ClassName:       element.Name,
				ClassSectionID:  element.SectionID,
				Kind:            SomMetaKindComplex,
				TypeName:        element.Name,
				DocComment:      element.Doc,
				ClassDocComment: element.Doc,
				MapsTo:          element.MapsTo,
				DetailedIn:      element.DetailedIn,
				Recursive:       elementRecursive,
				Children:        elementChildren,
			}
		}
	}

	var contentType *SomContentTypeMeta
	if field.ContentType != "" {
		description := ""
		if ctAnn := field.Annotation("ContentType"); ctAnn != nil {
			description = bridgeStr(ctAnn.Argument("description"))
		}
		contentType = &SomContentTypeMeta{Type: field.ContentType, Description: description}
	}

	commentText := ""
	if commentAnn := field.Annotation("Comment"); commentAnn != nil {
		commentText = bridgeStr(commentAnn.Argument("text"))
	}

	var form *SomFormMeta
	if field.Kind == SpecFieldKindForm {
		fields := make([]*SomFormFieldMeta, 0, len(field.FormFields))
		for i, ff := range field.FormFields {
			fields = append(fields, &SomFormFieldMeta{
				Name:        ff.Name,
				TypeName:    ff.Type,
				Description: ff.Label,
				Required:    ff.Required,
				Hint:        ff.Hint,
				Order:       i,
			})
		}
		form = &SomFormMeta{Fields: fields}
	}

	className := owner.Name
	docComment := field.Doc
	classDoc, mapsTo, detailedIn := "", "", ""
	classSectionID := ""
	if target != nil {
		className = target.Name
		if docComment == "" {
			docComment = target.Doc
		}
		classDoc = target.Doc
		classSectionID = target.SectionID
		mapsTo = target.MapsTo
		detailedIn = target.DetailedIn
	}
	typeName := field.Type
	if typeName == "" {
		typeName = field.ElementType
	}
	if typeName == "" {
		typeName = field.EnumType
	}
	if typeName == "" {
		typeName = "String"
	}

	return &SomMetaNode{
		ClassName:          className,
		MemberName:         field.Name,
		SectionID:          field.SectionID,
		ClassSectionID:     classSectionID,
		SectionIDPattern:   field.SectionIDPattern,
		Kind:               kind,
		TypeName:           typeName,
		SerializationOrder: field.SerializationOrder,
		Min:                field.Min,
		Unused:             field.Annotation("Unused") != nil,
		ContentType:        contentType,
		ContentHelp:        field.Help,
		Comment:            commentText,
		DocComment:         docComment,
		ClassDocComment:    classDoc,
		Form:               form,
		MapsTo:             mapsTo,
		DetailedIn:         detailedIn,
		SecondLevelIds:     bridgeSecondLevelIds(field.Annotations),
		Extra:              bridgeExtras(field.Annotations),
		Recursive:          recursive,
		Children:           children,
		ElementNode:        elementNode,
	}
}

func bridgeSecondLevelIds(annotations []*SpecAnnotation) []*SomSecondLevelId {
	var out []*SomSecondLevelId
	for _, a := range annotations {
		if a.Name != "SecondLevelSectionId" {
			continue
		}
		out = append(out, &SomSecondLevelId{
			DocumentClass: bridgeStr(a.Argument("documentClass")),
			ID:            bridgeStr(a.Argument("id")),
		})
	}
	return out
}

func bridgeExtras(annotations []*SpecAnnotation) []*SomMetaExtra {
	var out []*SomMetaExtra
	for _, a := range annotations {
		if slottedAnnotations[a.Name] {
			continue
		}
		args := a.Arguments
		if args == nil {
			args = map[string]interface{}{}
		}
		out = append(out, &SomMetaExtra{Annotation: a.Name, Args: args})
	}
	return out
}

File: tom_som_rust_runtime/src/spec_meta_bridge.rs
//! `spec_meta_bridge` — bridge from the meta-JSON class graph ([`SpecModel`])
//! to the canonical metadata tree ([`SomMetaTree`], DR1 §3.1), a faithful port
//! of the Go `spec_meta_bridge.go` (itself a port of
//! `tom_som_dart_runtime/lib/src/spec_meta_bridge.dart`).
//!
//! The generated facades (DR8) emit populated `SomMetaTree`s directly; every
//! consumer that only has the exported spec-model meta-data (the conformance
//! harness, tooling) builds its tree here. The expansion follows the same walk
//! `SpecReflection.resolve` performs — crucially, a node's
//! [`SomMetaNode::section_id`] is the **field-level** id exactly as exported
//! (`field.section_id`), so the tree's path grammar stays byte-compatible with
//! the paths a `SpecDocument` is keyed by. (DR1 §2.2's "field id, else
//! target-class id" resolution is applied at *export* time by the model
//! exporter / DR8 generator, not re-derived here.)
//!
//! Recursive class references become terminal re-entry nodes
//! ([`SomMetaNode::recursive`]), mirroring how the generated chains terminate.

use std::collections::HashSet;
use std::rc::Rc;

use crate::json::Json;
use crate::spec_meta::{
    SomContentTypeMeta, SomDocMeta, SomFormFieldMeta, SomFormMeta, SomMetaExtra, SomMetaNode,
    SomMetaTree, SomSecondLevelId, SOM_META_KIND_COMPLEX, SOM_META_KIND_SECTION,
};
use crate::spec_model::{
    SpecAnnotation, SpecClass, SpecField, SpecModel, SPEC_FIELD_KIND_COMPLEX,
    SPEC_FIELD_KIND_FORM, SPEC_FIELD_KIND_LIST, SPEC_FIELD_KIND_SECTION,
};

/// Reports whether an annotation name has a dedicated [`SomMetaNode`] slot;
/// everything else is captured losslessly into [`SomMetaNode::extra`].
fn is_slotted_annotation(name: &str) -> bool {
    matches!(
        name,
        "SectionId"
            | "SectionIdPattern"
            | "SerializationOrder"
            | "Min"
            | "Unused"
            | "ContentType"
            | "ContentHelp"
            | "Headline"
            | "Comment"
            | "Form"
            | "Document"
            | "MapsTo"
            | "DetailedIn"
            | "SecondLevelSectionId"
    )
}

/// Builds the wired [`SomMetaTree`] for one document root of `model`.
///
/// When `root_type` is `""` the model's first root is used; otherwise the root
/// is looked up via [`SpecModel::root_by_type`] (returning its error for an
/// unknown type). Children are ordered by `@SerializationOrder` (annotated
/// members first, then declaration order), which is the sibling emission order
/// of the hierarchical YAML format (DR1 §2.3).
pub fn build_som_meta_tree(model: &SpecModel, root_type: &str) -> Result<SomMetaTree, String> {
    let root = if root_type.is_empty() {
        match model.roots.first() {
            Some(r) => r,
            None => model.root_by_type("")?,
        }
    } else {
        model.root_by_type(root_type)?
    };
    let cls = model.class_named(&root.type_);

    let mut section_id = root.section_id.clone();
    let mut class_section_id = String::new();
    let mut doc_comment = root.doc.clone();
    let mut class_doc = String::new();
    let mut maps_to = String::new();
    let mut detailed_in = String::new();
    let mut root_headline = String::new();
    let mut annotations: &[SpecAnnotation] = &[];
    let mut children = Vec::new();
    if let Some(cls) = cls {
        if section_id.is_empty() {
            section_id = cls.section_id.clone();
        }
        class_section_id = cls.section_id.clone();
        if doc_comment.is_empty() {
            doc_comment = cls.doc.clone();
        }
        class_doc = cls.doc.clone();
        root_headline = cls.headline.clone();
        maps_to = cls.maps_to.clone();
        detailed_in = cls.detailed_in.clone();
        annotations = &cls.annotations;
        let mut stack = HashSet::new();
        stack.insert(root.type_.clone());
        children = bridge_child_nodes(model, cls, &mut stack);
    }
    let root_node = SomMetaNode {
        class_name: root.type_.clone(),
        section_id,
        class_section_id,
        kind: SOM_META_KIND_SECTION.to_string(),
        type_name: root.type_.clone(),
        headline: root_headline,
        doc_comment,
        class_doc_comment: class_doc,
        maps_to,
        detailed_in,
        document: Some(SomDocMeta {
            name: root.title.clone(),
            description: root.description.clone(),
            based_on: bridge_based_on(cls),
        }),
        second_level_ids: bridge_second_level_ids(annotations),
        extra: bridge_extras(annotations),
        children,
        ..SomMetaNode::default()
    };
    SomMetaTree::new(Rc::new(root_node))
}

fn bridge_based_on(cls: Option<&SpecClass>) -> Vec<String> {
    let Some(cls) = cls else {
        return Vec::new();
    };
    let Some(ann) = cls.annotation("Document") else {
        return Vec::new();
    };
    let Some(Json::Array(raw)) = ann.argument("basedOn") else {
        return Vec::new();
    };
    raw.iter().map(bridge_str).collect()
}

/// Coerces an annotation argument value to its string form (`""` for absent /
/// unrepresentable values, matching the other ports' `String(x || '')`).
fn bridge_str(v: &Json) -> String {
    match v {
        Json::Str(s) => s.clone(),
        Json::Int(i) => i.to_string(),
        Json::Float(f) => {
            if *f == (*f as i64) as f64 {
                (*f as i64).to_string()
            } else {
                String::new()
            }
        }
        Json::Bool(b) => {
            if *b {
                "true".to_string()
            } else {
                "false".to_string()
            }
        }
        _ => String::new(),
    }
}

fn bridge_child_nodes(
    model: &SpecModel,
    cls: &SpecClass,
    stack: &mut HashSet<String>,
) -> Vec<Rc<SomMetaNode>> {
    const FALLBACK: i64 = 1 << 30;
    let mut fields: Vec<(usize, &SpecField)> = cls.fields.iter().enumerate().collect();
    fields.sort_by_key(|(index, field)| (field.serialization_order.unwrap_or(FALLBACK), *index));
    fields
        .iter()
        .map(|(_, field)| Rc::new(bridge_field_node(model, cls, field, stack)))
        .collect()
}

fn bridge_field_node(
    model: &SpecModel,
    owner: &SpecClass,
    field: &SpecField,
    stack: &mut HashSet<String>,
) -> SomMetaNode {
    let kind = field.kind.clone(); // SPEC_FIELD_KIND_* values equal the SOM_META_KIND_* values.
    let mut target: Option<&SpecClass> = None;
    let mut recursive = false;
    let mut children = Vec::new();
    let mut element_node = None;

    if field.kind == SPEC_FIELD_KIND_COMPLEX || field.kind == SPEC_FIELD_KIND_SECTION {
        target = model.class_named(&field.type_);
        if let Some(t) = target {
            if stack.contains(&t.name) {
                recursive = true;
            } else {
                stack.insert(t.name.clone());
                children = bridge_child_nodes(model, t, stack);
                stack.remove(&t.name);
            }
        }
    } else if field.kind == SPEC_FIELD_KIND_LIST && field.element_is_complex {
        if let Some(element) = model.class_named(&field.element_type) {
            let mut element_children = Vec::new();
            let mut element_recursive = false;
            if stack.contains(&element.name) {
                element_recursive = true;
            } else {
                stack.insert(element.name.clone());
                element_children = bridge_child_nodes(model, element, stack);
                stack.remove(&element.name);
            }
            element_node = Some(Rc::new(SomMetaNode {
                class_name: element.name.clone(),
                class_section_id: element.section_id.clone(),
                kind: SOM_META_KIND_COMPLEX.to_string(),
                type_name: element.name.clone(),
                headline: element.headline.clone(),
                doc_comment: element.doc.clone(),
                class_doc_comment: element.doc.clone(),
                maps_to: element.maps_to.clone(),
                detailed_in: element.detailed_in.clone(),
                recursive: element_recursive,
                children: element_children,
                ..SomMetaNode::default()
            }));
        }
    }

    let content_type = if field.content_type.is_empty() {
        None
    } else {
        let description = field
            .annotation("ContentType")
            .and_then(|a| a.argument("description"))
            .map(bridge_str)
            .unwrap_or_default();
        Some(SomContentTypeMeta {
            type_: field.content_type.clone(),
            description,
        })
    };

    let comment_text = field
        .annotation("Comment")
        .and_then(|a| a.argument("text"))
        .map(bridge_str)
        .unwrap_or_default();

    let form = if field.kind == SPEC_FIELD_KIND_FORM {
        let fields = field
            .form_fields
            .iter()
            .enumerate()
            .map(|(i, ff)| SomFormFieldMeta {
                name: ff.name.clone(),
                type_name: ff.type_.clone(),
                description: ff.label.clone(),
                required: ff.required,
                hint: ff.hint.clone(),
                order: i as i64,
            })
            .collect();
        Some(SomFormMeta { fields })
    } else {
        None
    };

    let mut class_name = owner.name.clone();
    let mut class_section_id = String::new();
    let mut doc_comment = field.doc.clone();
    // YRD4: the field-level `@Headline` wins over the target class's.
    let mut headline = field.headline.clone();
    let mut class_doc = String::new();
    let mut maps_to = String::new();
    let mut detailed_in = String::new();
    if let Some(target) = target {
        class_name = target.name.clone();
        class_section_id = target.section_id.clone();
        if doc_comment.is_empty() {
            doc_comment = target.doc.clone();
        }
        if headline.is_empty() {
            headline = target.headline.clone();
        }
        class_doc = target.doc.clone();
        maps_to = target.maps_to.clone();
        detailed_in = target.detailed_in.clone();
    }
    let mut type_name = field.type_.clone();
    if type_name.is_empty() {
        type_name = field.element_type.clone();
    }
    if type_name.is_empty() {
        type_name = field.enum_type.clone();
    }
    if type_name.is_empty() {
        type_name = "String".to_string();
    }

    SomMetaNode {
        class_name,
        member_name: field.name.clone(),
        section_id: field.section_id.clone(),
        class_section_id,
        section_id_pattern: field.section_id_pattern.clone(),
        kind,
        type_name,
        serialization_order: field.serialization_order,
        min: field.min,
        unused: field.annotation("Unused").is_some(),
        content_type,
        content_help: field.help.clone(),
        headline,
        comment: comment_text,
        doc_comment,
        class_doc_comment: class_doc,
        form,
        maps_to,
        detailed_in,
        second_level_ids: bridge_second_level_ids(&field.annotations),
        extra: bridge_extras(&field.annotations),
        recursive,
        children,
        element_node,
        ..SomMetaNode::default()
    }
}

fn bridge_second_level_ids(annotations: &[SpecAnnotation]) -> Vec<SomSecondLevelId> {
    annotations
        .iter()
        .filter(|a| a.name == "SecondLevelSectionId")
        .map(|a| SomSecondLevelId {
            document_class: a.argument("documentClass").map(bridge_str).unwrap_or_default(),
            id: a.argument("id").map(bridge_str).unwrap_or_default(),
        })
        .collect()
}

fn bridge_extras(annotations: &[SpecAnnotation]) -> Vec<SomMetaExtra> {
    annotations
        .iter()
        .filter(|a| !is_slotted_annotation(&a.name))
        .map(|a| SomMetaExtra {
            annotation: a.name.clone(),
            args: a.arguments.clone(),
        })
        .collect()
}
